internal/config/resource_base.go

```diff
@@ -6,19 +6,53 @@ import (
 
 	"github.com/hashicorp/terraform-plugin-framework/datasource"
 	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/tfsdk"
+	"github.com/hashicorp/terraform-plugin-framework/types"
 )
 
 const (
 	errorConfigureSummary = "Unexpected Resource Configure Type"
 	errorConfigure        = "expected *MongoDBClient, got: %T. Please report this issue to the provider developers"
 )
 
+type ProviderMeta struct {
+	ModuleName     types.String `tfsdk:"module_name"`
+	ModuleVersion  types.String `tfsdk:"module_version"`
+	UserAgentExtra types.Map    `tfsdk:"user_agent_extra"`
+}
+
+type ImplementedResource interface {
+	resource.ResourceWithImportState
+	// Additional methods such as upgrade state & plan modifier are optional
+	SetClient(*MongoDBClient)
+	GetName() string
+}
+
+func AnalyticsResourceFunc(iResource resource.Resource) func() resource.Resource {
+	commonResource, ok := iResource.(ImplementedResource)
+	if !ok {
+		panic("resource %T didn't comply with the ImplementedResource interface")
+	}
+	a := func() resource.Resource {
+		return analyticsResource(commonResource)
+	}
+	return a
```

Comment:
```suggestion
	return func() resource.Resource {
		return analyticsResource(commonResource)
	}
```

nit: i prefer to return the variable instead of creating a "dummy" var

Reply:
ty changed in d5452ab66ba4ad2e8deb2fe6836299030ba5ebd1